// rs-restangular/dist/src/restangular.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var http_1 = require('@angular/http');
var config_1 = require('./config');
var path_1 = require('./path');
var Restangular = (function () {
    function Restangular(http) {
        this.http = http;
        this.config = new config_1.RestangularConfig();
        this.config.http = http;
        this.path = new path_1.RestangularPath(this.config);
    }
    Restangular.prototype.all = function (route) {
        return this.path.all(route);
    };
    Restangular.prototype.one = function (route, id) {
        return this.path.one(route, id);
    };
    Restangular = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [http_1.Http])
    ], Restangular);
    return Restangular;
}());
exports.Restangular = Restangular;
